src/image/TransmissionCurve.cc

```diff
@@ -0,0 +1,851 @@
+public:
+
+    std::shared_ptr<table::io::Persistable> read(
+        InputArchive const& archive,
+        CatalogVector const& catalogs
+    ) const override {
+        auto const& keys = TransformedPersistenceHelper::get();
+        LSST_ARCHIVE_ASSERT(catalogs.size() == 1u);
+        LSST_ARCHIVE_ASSERT(catalogs.front().getSchema() == keys.schema);
+        auto const & record = catalogs.front().front();
+        return std::make_shared<TransformedTransmissionCurve>(
+            archive.get<TransmissionCurve>(record.get(keys.nested)),
+            archive.get<geom::TransformPoint2ToPoint2>(record.get(keys.transform))
+        );
+    }
+
+    TransformedFactory(std::string const& name) : afw::table::io::PersistableFactory(name) {}
+};
+
+TransformedFactory transformedRegistration("TransformedTransmissionCurve");
+
+
+} // anonymous
+
+
+TransmissionCurve::SampleDef::SampleDef() :
+    min(-std::numeric_limits<double>::infinity()),
+    max(std::numeric_limits<double>::infinity()),
+    size(0)
+{}
+
+ndarray::Array<double,1,1> TransmissionCurve::SampleDef::makeArray() const {
+    ndarray::Array<double,1,1> result = ndarray::allocate(this->size);
+    double const step = (this->max - this->min)/(this->size - 1);
+    double current = this->min;
+    for (auto & element : result) {
+        element = current;
+        current += step;
+    }
+    result.back() = this->max; // make sure round-off error doesn't push us too far
+    return result;
+}
+
+std::shared_ptr<TransmissionCurve> TransmissionCurve::makeIdentity() {
+    return IdentityTransmissionCurve::get();
+}
+
+std::shared_ptr<TransmissionCurve> TransmissionCurve::makeConstant(
+    ndarray::Array<double const,1> const & throughput,
+    ndarray::Array<double const,1> const & wavelengths,
+    double throughputAtMin, double throughputAtMax
+) {
+    ::gsl_set_error_handler_off();
+    return std::make_shared<ConstantTransmissionCurve>(throughput, wavelengths,
+                                                       throughputAtMin, throughputAtMax);
+}
+
+std::shared_ptr<TransmissionCurve> TransmissionCurve::makeRadial(
+    ndarray::Array<double const,2> const & throughput,
+    ndarray::Array<double const,1> const & wavelengths,
+    ndarray::Array<double const,1> const & radii,
+    double throughputAtMin, double throughputAtMax
+) {
+    ::gsl_set_error_handler_off();
+    return std::make_shared<RadialTransmissionCurve>(throughput, wavelengths, radii,
+                                                     throughputAtMin, throughputAtMax);
+}
+
+std::shared_ptr<TransmissionCurve> TransmissionCurve::multiply(
+    std::shared_ptr<TransmissionCurve> a,
+    std::shared_ptr<TransmissionCurve> b
+) {
+    auto result = a->_multiplyImpl(a, b);
+    if (result == nullptr) {
+        result = b->_multiplyImpl(b, a);
+    }
+    if (result == nullptr) {
+        result = std::make_shared<ProductTransmissionCurve>(std::move(a), std::move(b));
```

Comment:
The constructor to `ProductTransmissionCurve` does not accept rvalue references; will these calls to `std::move` have any effect?

Reply:
My understanding is "yes", because the constructor does accept the `shared_ptr` arguments by value rather than by const ref.  So we move into the constructor argument, and then move out of the constructor argument into the data member, and in this case never have to invoke the copy constructor (which involves a mutex, while the `shared_ptr` move constructor does not).

A bit more context: I'm not sure passing `shared_ptr` by value and moving is actually better than passing it by const reference in general, but our coding standards actually say we should pass it by value (and have done so since before moving was an option, which was definitely less efficient than passing by const reference).  Now that moving is an option, passing `shared_ptr` by value and moving when you can be more efficient than passing `shared_ptr` by const reference, so I've switched to that rather than trying to fix our guidelines.

Reply:
Actually, I'm pretty sure you *do* copy the `shared_ptr` when initializing the argument by value. If you didn't, then this code would be unsafe, because you'd be clearing the original pointer held by the user.

Reply:
I think that's right in the worst case, but in the typical case the compiler's copy elision kicks in.  I *think* the point about pass-by-value + move being more efficient is really about copy elision into a by-value argument being more likely than copy elision out of a const reference argument.  But convincing myself all that was something I did a while ago and haven't thought deeply about since, so I should just redirect you to the source material:

https://web.archive.org/web/20140205194657/http://cpp-next.com/archive/2009/08/want-speed-pass-by-value/

It's a very interesting and controversial Dave Abrahams blog post (which has weirdly been almost lost to the internet, hence the wayback machine).

I'm not advocating pass-by-value as strongly as he does (and I think others have claimed that he was advocating it more than he actually did), but I think the case for `shared_ptr` is stronger than it is for other types, and given that it lets makes our current guidelines sensible in terms of performance, that's enough for me.

Reply:
Yes, I'm not objecting to the use of pass-by-value: smart pointers are supposed to behave like pointers, which means that, like iterators, they're optimized for by-value use (when's the last time you passed a C-style pointer by reference? 😉). My argument was that the `std::move` invocations were essentially dead code, and it was more a style comment than a correctness one.

Reply:
Gotcha.  I'm pretty confident that they're not dead code, but the line between premature optimization that obfuscates readability and a performance-motivated best practice is pretty thin here.